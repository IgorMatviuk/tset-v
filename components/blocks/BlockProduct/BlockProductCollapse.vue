<template>
  <div class="collapse-info">
    <button
      :aria-expanded="isVisible ? 'true' : 'false'"
      :class="{ 'collapse-info__title--active': isVisible }"
      class="collapse-info__title"
      @click="isVisible = !isVisible"
    >
      <slot name="title"></slot>
      <svg-icon
        class="collapse-info__icon"
        name="chevron"
        width="10"
        height="10"
      />
    </button>
    <BCollapse v-model="isVisible" accordion="collapseInfoAccordion">
      <div class="collapse-info__content">
        <div v-if="$slots.text" class="collapse-info__text">
          <slot name="text"></slot>
        </div>
        <div v-else-if="$slots.reviews" class="collapse-info__reviews">
          <slot name="reviews"></slot>
        </div>
      </div>
    </BCollapse>
  </div>
</template>

<script>
export default {
  data() {
    return {
      isVisible: false
    }
  }
}
</script>

<style lang="scss">
.collapse-info {
  border-bottom: 1px solid $light-1;

  &__title {
    @include reset-button-styles();

    display: flex;
    align-items: center;
    justify-content: space-between;
    width: 100%;
    padding-left: 2px;
    font-size: 12px;
    font-weight: 300;
    line-height: 40px;
    text-align: left;
    text-transform: uppercase;
  }

  &__title--active &__icon {
    transform: rotate(0);
  }

  &__icon {
    width: 10px;
    height: 10px;
    transition: transform 0.3s;
    transform: rotate(-90deg);
  }

  &__content {
    padding-bottom: 18px;
  }

  &__text {
    font-size: 13px;
    font-weight: 300;
    line-height: 16px;
  }

  &__reviews {
    padding-top: 17px;
    padding-bottom: 12px;
  }

  @include media-breakpoint-down(lg) {
    &__title {
      padding-left: 0;
    }

    &__text {
      font-size: 11px;
      line-height: 14px;
    }
  }

  @include media-breakpoint-down(sm) {
    &__title {
      font-size: 10px;
      line-height: 37.5px;
    }

    &__content {
      padding-bottom: 12px;
    }
  }
}
</style>
