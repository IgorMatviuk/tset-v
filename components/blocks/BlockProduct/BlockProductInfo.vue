<template>
  <div class="product-info">
    <div class="product-info__badge-wrapper">
      <div
        v-if="activeVariant.discount_price"
        class="product-info__badge product-info__badge--sale"
      >
        Sale
      </div>
      <div
        v-if="activeVariant.is_new"
        class="product-info__badge product-info__badge--new"
      >
        New
      </div>
    </div>
    <div class="product-info__toggle">
      <NuxtLink :to="$i18n.path(links.prev)" class="product-info__toggle-link">
        <svg-icon
          class="product-info__toggle-icon"
          name="chevron"
          width="10"
          height="10"
        />
        {{ $t('productPage.prev') }}
      </NuxtLink>
      <NuxtLink :to="$i18n.path(links.next)" class="product-info__toggle-link">
        {{ $t('productPage.next') }}
        <svg-icon
          class="product-info__toggle-icon product-info__toggle-icon--reverse"
          name="chevron"
          width="10"
          height="10"
        />
      </NuxtLink>
    </div>
    <div class="product-info__name">
      {{ activeVariant.name }}
    </div>
    <div class="product-info__code">
      {{ $t('productPage.sku') }}: {{ variantByOptions.sku }}
    </div>
    <div class="product-info__price-wrapper">
      <template v-if="variantByOptions.discount_price">
        <div class="product-info__old-price">
          {{ variantByOptions.formatted_price }}
        </div>
        <div class="product-info__price">
          {{ variantByOptions.formatted_discount_price }}
        </div>
      </template>
      <template v-else>
        <div class="product-info__price">
          {{ variantByOptions.formatted_price }}
        </div>
      </template>
    </div>
    <div class="product-info__sizes">
      <button
        v-for="size in sizeOptions.items"
        :key="size.id"
        :class="{
          'product-info__size-item--active': selectedSize
            ? selectedSize.id === size.id
            : initialSelectedSize.id === size.id
        }"
        :disabled="getIsSizeOptionDisabled(size.variants)"
        class="product-info__size-item"
        @click="onSelectSize(size)"
      >
        {{ size.label }}
      </button>
    </div>
    <button class="product-info__size-table" @click="toggleFloatPanel(true)">
      {{ $t('floatPanel.title') }}
    </button>
    <div class="product-info__colors">
      <div class="product-info__active-color">
        {{ activeColorLabel }}
      </div>
      <div
        v-for="color in colorOptions.items"
        :key="color.id"
        class="product-info__colors-item-wrapper"
      >
        <!-- <span
          v-if="getIsColorOptionDisabled(color.variants)"
          :style="`background-color: ${color.swatch_value}`"
          :class="{
            'product-info__colors-item--active': selectedColor
              ? selectedColor.id === color.id
              : initialSelectedColor.id === color.id
          }"
          class="product-info__colors-item product-info__colors-item--disabled"
        ></span> -->
        <NuxtLink
          :to="getVariationUrl(color)"
          :style="`background-color: ${color.swatch_value}`"
          :class="{
            'product-info__colors-item--active': selectedColor
              ? selectedColor.id === color.id
              : initialSelectedColor.id === color.id,
            'product-info__colors-item--disabled': getIsColorOptionDisabled(
              color.variants
            )
          }"
          class="product-info__colors-item"
        />
      </div>
    </div>
    <!-- TODO: dynamics tags -->
    <div class="product-info__tag">
      <svg-icon
        name="free-shipping"
        class="product-info__tag-icon"
        width="12"
        height="12"
      />
      <div
        class="product-info__tag-content"
        v-html="$t('productTags.delivery')"
      ></div>
    </div>
    <div class="product-info__tag">
      <svg-icon
        name="exchange"
        class="product-info__tag-icon"
        width="12"
        height="12"
      />
      <div
        class="product-info__tag-content"
        v-html="$t('productTags.return')"
      ></div>
    </div>
    <!-- TODO: stock modal, add to cart -->
    <!-- v-if="variant.in_stock" -->
    <BButton
      block
      variant="primary"
      class="product-info__add"
      :class="{ 'product-info__add--disabled': !canBuyProduct }"
      @click="onAddToCart"
    >
      {{ $t('cart.add') }}
    </BButton>
    <!-- <BButton
      v-else
      block
      variant="outline-light"
      class="product-info__add product-info__add--no-stock"
      @click="onStockSubscribe"
    >
      сообщить о появлении
    </BButton> -->
    <BlockProductSpecs :items="variantSpecItems" />
    <BlockProductDescription
      :content="item.attribute_family.description.short_description.value"
    />
    <BlockProductAdditionalInfo :items="item.attribute_family.info" />
    <BlockProductReviews v-if="item.reviews.length" :items="item.reviews" />
  </div>
</template>

<script>
import { mapState } from 'vuex'

import BlockProductSpecs from './BlockProductSpecs'
import BlockProductDescription from './BlockProductDescription'
import BlockProductAdditionalInfo from './BlockProductAdditionalInfo'
import BlockProductReviews from './BlockProductReviews'

import productVariants from '@/mixins/product-variants'
import addToCart from '@/mixins/add-to-cart'

import { mutationTypes as productMutationTypes } from '@/store/product'

export default {
  components: {
    BlockProductSpecs,
    BlockProductDescription,
    BlockProductAdditionalInfo,
    BlockProductReviews
  },
  mixins: [productVariants, addToCart],
  props: {
    item: {
      type: Object,
      required: true
    },
    links: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      isFloatPanelVisible: false
    }
  },
  computed: {
    ...mapState({
      selectedSize: (state) => state.product.selectedSize,
      selectedColor: (state) => state.product.selectedColor,
      activeVariant: (state) => state.product.activeVariant
    }),
    variantByOptions() {
      const color = this.selectedColor || this.initialSelectedColor
      const size = this.selectedSize || this.initialSelectedSize

      return this.getVariantBySelectedOptions(color, size)
    },
    variantSpecItems() {
      const color = this.selectedColor || this.initialSelectedColor
      const size = this.selectedSize || this.initialSelectedSize

      const byId = (id) => (x) => x.value === id

      const variant = this.item.variants.find((variant) => {
        const vColor = variant.attributes.find(byId(color.id))
        const vSize = variant.attributes.find(byId(size.id))

        return (Boolean(vColor) && Boolean(vSize)) || Boolean(vColor)
      })

      const spec = Object.keys(variant.attribute_family.spec).reduce(
        (acc, key) => {
          const spec = variant.attribute_family.spec[key].value
            ? variant.attribute_family.spec[key]
            : this.item.attribute_family.spec[key]

          acc[key] = spec

          return acc
        },
        {}
      )

      return spec
    },
    canBuyProduct() {
      const color = this.selectedColor || this.initialSelectedColor
      const isColorDisabled = this.getIsColorOptionDisabled(color.variants)

      return !isColorDisabled
    },
    activeColorLabel() {
      const color = this.selectedColor || this.initialSelectedColor
      return color.label
    }
  },
  methods: {
    getVariationUrl(color) {
      const size = this.selectedSize || this.initialSelectedSize

      const variant = this.getVariantBySelectedOptions(color, size)

      return this.$i18n.path(variant.url_key)
    },
    getIsSizeOptionDisabled(variants) {
      const intersection = variants.filter((variant) =>
        this.activeColorVariants.includes(variant)
      )

      return !intersection.length
    },
    getIsColorOptionDisabled(variants) {
      const intersection = variants.filter((variant) =>
        this.activeSizeVariants.includes(variant)
      )

      return !intersection.length
    },
    onSelectSize(size) {
      this.activeSizeVariants = size.variants

      this.$store.commit(
        `product/${productMutationTypes.SET_SELECTED_SIZE}`,
        size
      )
    },
    onAddToCart() {
      const color = this.selectedColor || this.initialSelectedColor
      const size = this.selectedSize || this.initialSelectedSize

      const variant = this.getVariantBySelectedOptions(color, size)

      this.addToCart(variant.id)
    },
    onStockSubscribe() {
      this.$bus.$emit('stock-subscribe')
    },
    toggleFloatPanel(status) {
      this.$bus.$emit('float-panel:toggle', status)
    }
  }
}
</script>

<style lang="scss">
.product-info {
  display: flex;
  flex-wrap: wrap;
  align-items: flex-start;

  &__badge-wrapper {
    display: flex;
    align-items: flex-start;
  }

  &__badge {
    width: 60px;
    height: 25px;
    font-size: 12px;
    font-weight: 500;
    line-height: 23px;
    color: $white;
    text-align: center;
    text-transform: uppercase;
    border: 1px solid rgba($color: $black, $alpha: 0.15);

    &:not(:last-of-type) {
      margin-right: 10px;
    }

    &--new {
      color: $black;
      background-color: $white;
    }

    &--sale {
      background-color: $primary;
    }
  }

  &__toggle {
    position: absolute;
    top: 0;
    right: 0;
    line-height: 0.4;
  }

  &__toggle-link {
    display: inline-block;
    font-size: 9px;
    font-weight: 300;
    line-height: 11px;

    &:not(:last-of-type)::after {
      margin-left: 4px;
      content: '/';
    }
  }

  &__toggle-icon {
    display: none;
  }

  &__name {
    width: 100%;
    padding-right: 15px;
    margin-top: 24px;
    font-size: 13px;
    font-weight: 300;
    line-height: 16px;
    text-transform: uppercase;
    letter-spacing: 0.07em;
  }

  &__code {
    width: 100%;
    margin-top: 5px;
    font-size: 12px;
    font-weight: normal;
    line-height: 15px;
    color: $gray-dark-1;
  }

  &__price-wrapper {
    display: flex;
    align-items: baseline;
    width: 100%;
    margin-top: 12px;
  }

  &__old-price {
    margin-right: 13px;
    font-size: 14px;
    font-weight: 300;
    line-height: 30px;
    color: $text-color;
    text-decoration-line: line-through;
  }

  &__price {
    font-size: 16px;
    line-height: 30px;
    color: $primary;
  }

  &__sizes {
    display: flex;
    width: 100%;
    margin-top: 16px;
  }

  &__size-item {
    @include reset-button-styles();

    width: 45px;
    height: 25px;
    font-size: 13px;
    font-weight: 300;
    line-height: 23px;
    text-align: center;
    border: 1px solid $light-1;
    transition: color 0.3s, background-color 0.3s, border-color 0.3s;

    &:not(:last-of-type) {
      margin-right: 10px;
    }

    &--active {
      color: $white;
      cursor: default;
      background-color: $primary;
      border-color: $primary;
    }

    &:disabled {
      cursor: default;
    }

    &:not(:disabled) {
      @include hover-focus-active {
        color: $white;
        background-color: $primary;
        border-color: $primary;
      }
    }
  }

  &__size-table {
    @include reset-button-styles();

    @include hover-focus-active() {
      color: $primary;
    }

    margin-top: 15px;
    font-size: 12px;
    font-weight: 300;
    line-height: 15px;
    text-transform: uppercase;
    letter-spacing: 0.1em;
    text-decoration-line: underline;
    transition: color 0.3s;
  }

  &__colors {
    display: flex;
    flex-wrap: wrap;
    align-content: flex-start;
    width: 100%;
    margin-top: 28px;
    margin-bottom: 31px;
  }

  &__colors-item-wrapper {
    &:not(:last-of-type) {
      margin-right: 13px;
    }
  }

  &__colors-item {
    position: relative;
    display: block;
    width: 25px;
    height: 16px;
    border: 1px solid rgba($color: $black, $alpha: 0.1);

    &--active {
      pointer-events: none;
      box-shadow: 0 0 0 2px $white, 0 0 0 3px $gray-dark-1;
    }

    &--disabled {
      opacity: 0.5;
      transition: opacity 0.3s;

      &:hover {
        opacity: 1;
      }

      &::after {
        position: absolute;
        top: 50%;
        right: 0;
        left: 0;
        height: 1px;
        content: '';
        background-color: $white;
        transform: rotate(-45deg);
      }
    }
  }

  &__active-color {
    order: 1;
    width: 100%;
    margin-top: 10px;
    font-size: 12px;
    font-weight: 300;
    line-height: 15px;
    color: $text-color;
    letter-spacing: 0.07em;
  }

  &__add {
    margin-top: 22px;
    font-size: 14px;

    &--no-stock {
      border-color: #ccc;

      @include hover-focus-active() {
        border-color: #ccc !important;
      }

      &:focus {
        box-shadow: 0 0 0 0.2rem rgba($color: #ccc, $alpha: 0.5);
      }
    }

    &--disabled {
      pointer-events: none;
      opacity: 0.5;
    }
  }

  &__collapse {
    width: 100%;
  }

  &__review {
    &:not(:last-of-type) {
      margin-bottom: 22px;
    }
  }

  &__tag {
    display: flex;
    align-items: center;
    width: 100%;
    margin-bottom: 9px;
  }

  &__tag-icon {
    width: 12px;
    height: 12px;
    margin-right: 9px;
    color: $primary-light-1;
  }

  &__tag-content {
    font-size: 12px;
    font-weight: 300;
    line-height: 15px;
    color: $black;

    b {
      font-weight: 500;
    }

    a {
      font-weight: 500;
      text-decoration-line: underline;
      transition: color 0.3s;
    }
  }

  @include media-breakpoint-down(xl) {
    &__colors {
      margin-bottom: 26px;
    }

    &__add {
      margin-top: 16px;
    }

    &__tag-icon {
      margin-right: 13px;
    }
  }

  @include media-breakpoint-down(lg) {
    &__name {
      margin-top: 23px;
      font-size: 12px;
    }

    &__code {
      margin-top: 8px;
    }

    &__price-wrapper {
      margin-top: 10px;
    }

    &__old-price {
      margin-right: 7px;
      font-size: 12px;
    }

    &__price {
      font-size: 14px;
    }

    &__sizes {
      margin-top: 9px;
    }

    &__colors {
      margin-top: 21px;
      margin-bottom: 20px;
    }

    &__tag {
      margin-bottom: 10px;
    }

    &__tag-icon {
      margin-right: 9px;
    }

    &__tag-content {
      font-size: 11px;
      line-height: 14px;
    }

    &__add {
      margin-top: 12px;
    }
  }

  @include media-breakpoint-down(sm) {
    &__badge-wrapper {
      position: absolute;
      top: 39px;
      left: 15px;
      z-index: 1;
    }

    &__badge {
      width: 48px;
      height: 20px;
      font-size: 10px;
      line-height: 20px;
    }

    &__toggle {
      top: 0;
      right: 13px;
      left: 13px;
      display: flex;
      justify-content: space-between;
    }

    &__toggle-link {
      color: $text-color;
      letter-spacing: 0.07em;

      &:not(:last-of-type)::after {
        display: none;
      }
    }

    &__toggle-icon {
      display: inline-block;
      margin-bottom: 1px;
      color: $gray-dark-2;
      transform: rotate(90deg);

      &--reverse {
        transform: rotate(-90deg);
      }
    }

    &__colors {
      justify-content: center;
      order: -1;
      margin-top: 10px;
    }

    &__active-color {
      order: -1;
      margin-top: 0;
      margin-bottom: 10px;
      font-size: 9px;
      line-height: 11px;
      text-align: center;
    }

    &__name {
      max-width: 85%;
      padding-right: 0;
      margin-top: 3px;
      margin-right: auto;
      margin-left: auto;
      line-height: 15px;
      text-align: center;
    }

    &__code {
      margin-top: 4px;
      font-size: 11px;
      line-height: 14px;
      text-align: center;
    }

    &__price-wrapper {
      justify-content: center;
      padding-left: 7px;
      margin-top: 14px;
    }

    &__old-price {
      margin-right: 10px;
    }

    &__sizes {
      justify-content: center;
    }

    &__size-table {
      margin-top: 7px;
      margin-right: auto;
      margin-bottom: 25px;
      margin-left: auto;
      font-size: 9px;
      line-height: 30px;
      letter-spacing: 0.1em;
    }

    &__add {
      margin-top: 17px;
      margin-right: 15px;
      margin-left: 15px;
    }

    &__tag {
      justify-content: center;
      padding-right: 25px;
      margin-bottom: 14px;
    }

    &__tag-icon {
      margin-right: 14px;
    }
  }
}
</style>
