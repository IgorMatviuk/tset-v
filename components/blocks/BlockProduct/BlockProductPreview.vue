<template>
  <div
    class="product-preview-card"
    @mouseover="activateHoverZoom"
    @click="showImageViewer"
  >
    <template v-if="isImagePreview">
      <ImageLoader ref="imageWrapper" class="product-preview-card__img-wrapper">
        <picture>
          <source
            :data-srcset="item.preview_lg | webp"
            srcset="/img/stub.png"
            media="(max-width: 1439px)"
            type="image/webp"
          />
          <source
            :data-srcset="item.preview_lg"
            srcset="/img/stub.png"
            media="(max-width: 1439px)"
            type="image/jpeg"
          />
          <source
            :data-srcset="item.preview_xl | webp"
            srcset="/img/stub.png"
            type="image/webp"
          />
          <img
            :data-src="item.preview_xl"
            :title="item.title"
            :alt="item.alt"
            src="/img/stub.png"
            class="product-preview-card__img lazyload"
            loading="lazy"
            data-object-fit="cover"
          />
        </picture>
        <ZoomedImage v-if="hoverZoomIsActive" :image-url="item.preview_xl" />
      </ImageLoader>
    </template>
    <template v-else>
      <BlockProductPreviewVideo :video="item" />
    </template>
  </div>
</template>

<script>
import PinchZoom from 'pinch-zoom-js'

import BlockProductPreviewVideo from './BlockProductPreviewVideo'

import ImageLoader from '@/components/shared/ImageLoader'
import ZoomedImage from '@/components/shared/ZoomedImage'

import { mutationTypes as productMutationsTypes } from '@/store/product'

import imageNextGen from '@/mixins/image-next-gen'

export default {
  components: {
    ImageLoader,
    ZoomedImage,
    BlockProductPreviewVideo
  },
  mixins: [imageNextGen],
  props: {
    item: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      hoverZoomIsActive: false,
      pinchZoom: null
    }
  },
  computed: {
    isImagePreview() {
      return this.item.preview_type === 'image'
    }
  },
  mounted() {
    if (this.$refs.imageWrapper) {
      this.setupPinchZoom()
    }

    window.addEventListener('resize', this.destroyHoverZoom)
  },
  beforeDestroy() {
    if (this.pinchZoom) {
      this.pinchZoom.disable()
      this.pinchZoom = null
    }

    window.removeEventListener('resize', this.destroyHoverZoom)
  },
  methods: {
    setupPinchZoom() {
      const element = this.$refs.imageWrapper.$el
      const mqlLg = window.matchMedia('(max-width: 1199.98px)')

      /* istanbul ignore else */
      if (element && mqlLg.matches) {
        this.pinchZoom = new PinchZoom(element, { draggableUnzoomed: false })
      }
    },
    destroyHoverZoom() {
      this.hoverZoomIsActive = false
    },
    activateHoverZoom() {
      const mqlXl = window.matchMedia('(min-width: 1200px)')

      /* istanbul ignore else */
      if (mqlXl.matches) {
        this.hoverZoomIsActive = true
      }
    },
    showImageViewer() {
      /* istanbul ignore else */
      if (this.isImagePreview) {
        this.$store.commit(
          `product/${productMutationsTypes.SET_ACTIVE_PREVIEW}`,
          this.item
        )
        this.$store.commit(
          `product/${productMutationsTypes.SET_IMAGE_VIEWER_VISIBILITY_TOGGLE}`,
          true
        )
      }
    }
  }
}
</script>

<style lang="scss">
.product-preview-card {
  position: relative;
  height: 1px;
  padding-bottom: $preview-aspect-ratio;
  overflow: hidden;
  transform: translateZ(0);
  backface-visibility: hidden;

  &__img {
    width: 100%;
    height: 100%;
    object-fit: cover;
  }

  .pinch-zoom-container {
    position: absolute !important;
    width: 100%;
  }

  .zoomed {
    opacity: 0;
    transition: opacity 0.5s;
  }

  &:hover .zoomed {
    opacity: 1;
  }

  @include media-breakpoint-down(sm) {
    position: relative;
    width: 100%;
    height: 0;

    &__img-wrapper {
      width: 100%;
      height: 100%;
    }
  }
}
</style>
