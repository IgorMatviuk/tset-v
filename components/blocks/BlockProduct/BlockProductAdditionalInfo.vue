<template>
  <div class="product-info__additional-info">
    <BlockProductCollapse
      v-for="content in items"
      :key="content.code"
      class="product-info__collapse"
    >
      <template v-slot:title>
        <span>{{ content.name }}</span>
      </template>
      <template v-slot:text>
        <span>{{ content.value }}</span>
      </template>
    </BlockProductCollapse>
  </div>
</template>

<script>
import BlockProductCollapse from './BlockProductCollapse'

export default {
  components: {
    BlockProductCollapse
  },
  props: {
    items: {
      type: Object,
      required: true
    }
  }
}
</script>

<style lang="scss">
.product-info__additional-info {
  width: 100%;
}
</style>
