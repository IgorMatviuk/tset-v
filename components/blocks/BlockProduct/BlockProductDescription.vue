<template>
  <BlockProductCollapse class="product-info__collapse">
    <template v-slot:title>
      <span>{{ $t('productPage.info.description') }}</span>
    </template>
    <template v-slot:text>
      <div v-html="content"></div>
    </template>
  </BlockProductCollapse>
</template>

<script>
import BlockProductCollapse from './BlockProductCollapse'

export default {
  components: {
    BlockProductCollapse
  },
  props: {
    content: {
      type: String,
      required: true
    }
  }
}
</script>

<style lang="scss">
.collapse-info__text {
  p {
    &:only-child {
      margin-bottom: 0;
    }
  }
}
</style>
