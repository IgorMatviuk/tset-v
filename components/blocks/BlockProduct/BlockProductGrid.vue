<template>
  <div class="product-grid">
    <div
      v-for="(preview, index) in formattedPreviews"
      :key="index"
      class="product-grid__item"
    >
      <BlockProductPreview :item="preview" />
    </div>
  </div>
</template>

<script>
import BlockProductPreview from './BlockProductPreview'

import formattedPreviews from '@/mixins/formatted-previews'

export default {
  components: {
    BlockProductPreview
  },
  mixins: [formattedPreviews],
  props: {
    previews: {
      type: Array,
      required: true
    }
  }
}
</script>

<style lang="scss">
.product-grid {
  display: none;

  &__item {
    &:not(:last-of-type) {
      margin-bottom: 6px;
    }
  }

  @include media-breakpoint-up(md) {
    display: block;
  }

  @include media-breakpoint-up(lg) {
    display: flex;
    flex-wrap: wrap;
    margin-left: -2.5px;

    &__item {
      flex: 0 0 50%;
      max-width: 50%;
      padding-right: 2.5px;
      padding-left: 2.5px;
    }
  }

  @include media-breakpoint-up(xl) {
    margin-left: -4px;

    &__item {
      padding-right: 4px;
      padding-left: 4px;

      &:not(:last-of-type) {
        margin-bottom: 8px;
      }
    }
  }
}
</style>
