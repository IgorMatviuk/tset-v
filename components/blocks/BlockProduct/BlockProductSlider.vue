<template>
  <div class="product-slider">
    <div
      v-swiper:productSlider="swiperOptions"
      :cleanup-styles-on-destroy="false"
      class="product-slider__container"
    >
      <div class="swiper-wrapper product-slider__wrapper">
        <div
          v-for="(preview, index) in formattedPreviews"
          :key="index"
          class="swiper-slide product-slider__slide"
        >
          <BlockProductPreview ref="previewItem" :item="preview" />
        </div>
      </div>
      <SliderPagination class="product-slider__pagination" />
    </div>
  </div>
</template>

<script>
import BlockProductPreview from './BlockProductPreview'
import SliderPagination from '@/components/shared/SliderPagination'

import formattedPreviews from '@/mixins/formatted-previews'

export default {
  components: {
    BlockProductPreview,
    SliderPagination
  },
  mixins: [formattedPreviews],
  props: {
    previews: {
      type: Array,
      required: true
    }
  },
  data() {
    return {
      swiperOptions: {
        pagination: {
          el: '.product-slider__pagination',
          clickable: true
        },
        on: {
          slideChange: this.onSlideChange
        }
      }
    }
  },
  methods: {
    onSlideChange() {
      const video = this.$el.querySelector('video')

      if (this.productSlider.activeIndex === 1) {
        this.toggleVideoPlayState(video, 'play')
      } else {
        this.toggleVideoPlayState(video, 'pause')
      }
    },
    toggleVideoPlayState(video, nextState) {
      const isPlaying = !video.paused || !video.ended

      if (nextState === 'play') {
        video.play()
      } else if (isPlaying) {
        video.pause()
      }
    }
  }
}
</script>

<style lang="scss">
.product-slider {
  &__pagination {
    .swiper-pagination-bullet {
      bottom: 7px !important;
      width: 8px;
      height: 8px;
      margin: 0 5.5px !important;
    }
  }

  @include media-breakpoint-up(md) {
    display: none;

    &__pagination {
      display: none;
    }
  }
}
</style>
