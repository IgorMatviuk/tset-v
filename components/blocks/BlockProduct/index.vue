<template>
  <div class="product">
    <div class="product__container">
      <transition name="fade-slider">
        <BlockProductSlider
          v-if="isMobile"
          :previews="previews"
          class="product__slider"
        />
        <BlockProductGrid v-else :previews="previews" class="product__slider" />
      </transition>
      <BlockProductInfo
        :item="product"
        :links="links"
        :page="'product'"
        class="product__info"
      />
    </div>
  </div>
</template>

<script>
import BlockProductSlider from './BlockProductSlider'
import BlockProductGrid from './BlockProductGrid'
import BlockProductInfo from './BlockProductInfo'

export default {
  components: {
    BlockProductSlider,
    BlockProductGrid,
    BlockProductInfo
  },
  props: {
    product: {
      type: Object,
      required: true
    },
    links: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      isMobile: false,
      mqlMd: null
    }
  },
  computed: {
    previews() {
      // if active variant
      // else product
      const variant = this.product.variants.find(
        (variant) => variant.url_key === this.$route.params.pathMatch
      )

      if (variant) {
        return variant.previews
      } else if (this.product.variants.length) {
        return this.product.variants[0].previews
      } else {
        return this.product.previews
      }
    }
  },
  created() {
    if (this.$device.isMobile) {
      this.isMobile = true
    }
  },
  mounted() {
    this.mqlMd = window.matchMedia('(min-width: 768px)')

    this.mqlMd.addListener(this.checkDevice)
  },
  beforeDestroy() {
    this.mqlMd.removeListener(this.checkDevice)
  },
  methods: {
    checkDevice() {
      if (this.mqlMd.matches) {
        this.isMobile = false
      } else {
        this.isMobile = true
      }
    }
  }
}
</script>

<style lang="scss">
.product {
  position: relative;
  z-index: 11;

  &__container {
    display: flex;
    align-items: flex-start;
  }

  &__slider {
    position: sticky;
    top: 80px;
    width: 64.5%;
    margin-right: 95px;
  }

  &__info {
    position: sticky;
    top: 80px;
    width: 330px;
  }

  @include media-breakpoint-down(xl) {
    &__slider {
      width: 63.55%;
      margin-right: 68px;
    }
  }

  @include media-breakpoint-down(lg) {
    &__slider {
      width: 63.05%;
      margin-right: 31px;
    }

    &__info {
      width: 290px;
    }
  }

  @include media-breakpoint-down(md) {
    &__slider {
      width: 55.9%;
    }

    &__info {
      width: 270px;
    }
  }

  @include media-breakpoint-down(sm) {
    &__container {
      position: relative;
      display: block;
      padding-top: 24px;
    }

    &__slider {
      position: static;
      width: 100%;
      margin-right: 0;
    }

    &__info {
      position: static;
      width: 100%;
      max-width: 540px;
      padding-right: 15px;
      padding-left: 15px;
      margin-right: auto;
      margin-left: auto;
    }
  }
}
</style>
