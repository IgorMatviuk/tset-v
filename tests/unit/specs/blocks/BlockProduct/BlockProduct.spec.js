import { shallowMount } from '@vue/test-utils'

import BlockProduct from '@/components/blocks/BlockProduct'
import BlockProductGrid from '@/components/blocks/BlockProduct/BlockProductGrid'
import BlockProductSlider from '@/components/blocks/BlockProduct/BlockProductSlider'

import createFactory from '@/api/helpers/factory'
import productAttrs from '@/api/factories/product'

describe('BlockProduct', () => {
  let regularProduct

  function createMatchMediaMock(matches) {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches,
        media: query,
        addListener: jest.fn().mockImplementation((callback) => callback()),
        removeListener: jest.fn().mockImplementation((callback) => callback()),
        addEventListener: jest.fn().mockImplementation((callback) => callback())
      }))
    })
  }

  function createBlockProduct(mockup) {
    const wrapper = shallowMount(BlockProduct, {
      propsData: {
        product: regularProduct,
        links: {
          prev: 'prev-product-slug',
          next: 'next-product-slug'
        }
      },
      mocks: {
        $route: {
          params: {
            pathMatch: regularProduct.variants[0].url_key
          }
        },
        ...mockup
      }
    })

    return wrapper
  }

  beforeEach(() => {
    const products = createFactory(productAttrs, 12)
    regularProduct = products.find((product) => product.in_stock)
  })

  test('renders component without slider', () => {
    createMatchMediaMock(true)

    const wrapper = createBlockProduct({ $device: { isMobile: false } })

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.findComponent(BlockProductGrid).element).toBeVisible()
  })

  test('renders component without slider', () => {
    createMatchMediaMock(false)

    const wrapper = createBlockProduct({ $device: { isMobile: true } })

    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.findComponent(BlockProductSlider).element).toBeVisible()
  })

  test('destroy product block', () => {
    createMatchMediaMock(false)

    const wrapper = createBlockProduct({ $device: { isMobile: true } })

    wrapper.destroy()
  })
})
