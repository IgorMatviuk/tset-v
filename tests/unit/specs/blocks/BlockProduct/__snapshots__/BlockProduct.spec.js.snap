// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockProduct renders component without slider 1`] = `
<div
  class="product"
>
  <div
    class="product__container"
  >
    <transition-stub
      name="fade-slider"
    >
      <blockproductgrid-stub
        class="product__slider"
        previews="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
      />
    </transition-stub>
     
    <blockproductinfo-stub
      class="product__info"
      item="[object Object]"
      links="[object Object]"
      page="product"
    />
  </div>
</div>
`;

exports[`BlockProduct renders component without slider 2`] = `
<div
  class="product"
>
  <div
    class="product__container"
  >
    <transition-stub
      name="fade-slider"
    >
      <blockproductslider-stub
        class="product__slider"
        previews="[object Object],[object Object],[object Object],[object Object],[object Object],[object Object]"
      />
    </transition-stub>
     
    <blockproductinfo-stub
      class="product__info"
      item="[object Object]"
      links="[object Object]"
      page="product"
    />
  </div>
</div>
`;
